### DISCUSSION TITLE
What dependencies are required for `--engine llvm` ?

### DISCUSSION THREAD

**Discussion by juliusl** [xd009642/tarpaulin#1627] [Q&A] [answered]
So on my local dev machine I'm able to get `--engine llvm` to work perfectly, but I'm having issues in my CI pipeline.  Specifically in a CentOS Distro, which packages are required to be installed? 

I tried installing `llvm16 llvm16-devel clang16 clang16-devel`, but I keep getting this error:

`/__w/1/s/target/debug/deps/runplat_macros-2cec8885168ed29a: error while loading shared libraries: libstd-b420905fdf897c89.so: cannot open shared object file: No such file or directory`

**Comment by xd009642** [comment, marked as answer]
So libstd is libc, so you'll require a c standard library. If you don't want to use coveralls you can build without default features and that'll remove the openssl requirements. Apart from that I think the only other requirement would be a rust tool chain (cargo etc). Managing the rust install via rustup may also be nicer as tarpaulin can make use of some of the rustup environment to make it's life easier

**Reply by xd009642** [reply]
Hopefully there's nothing else hiding in the dependency tree but if there is you can do cargo tree on the project and check the dependency tree (I can do this when I'm on Pc for you as well)

**Reply by juliusl** [reply]
>So libstd is libc, so you'll require a c standard library

Interesting, I'm pretty sure that's there because `build-essential` is there. I also require openssl so I have that already and it seems to be fine. Rust toolchain is there as well, but due to policy I'm not allowed to use rustup so we have to manage our rust environment through an in-house tool. It might be something rustup is doing because on my environments where it's working I do use rustup.

**Reply by xd009642** [reply]
hmmm you can build with musl if that's easier as well :thinking:

**Reply by juliusl** [reply]
So found something interesting out,

On the environments that work,
```
#  Works - https://github.com/microsoft/AzureLinux
❯ ldconfig -p | grep libstd                                                            
        libstdc++.so.6 (libc6,x86-64) => /lib/libstdc++.so.6
        libstdc++.so (libc6,x86-64) => /lib/libstdc++.so
        libstd-78b5dabc3cec94b9.so (libc6,x86-64) => /lib/libstd-78b5dabc3cec94b9.so

# Works - Ubuntu 22.04
❯ ldconfig -p | grep libstd                                                                                                                                                                                                
        libstdc++.so.6 (libc6,x86-64) => /lib/x86_64-linux-gnu/libstdc++.so.6
```

And in the CI version,
```
# Fails - https://github.com/microsoft/AzureLinux
❯ ldconfig -p | grep libstd
	libstdc++.so.6 (libc6,x86-64) => /lib/libstdc++.so.6
	libstdc++.so (libc6,x86-64) => /lib/libstdc++.so

2024-09-24T20:13:04.9263330Z /__w/1/s/target/debug/deps/runplat_macros-d78adfff56e6431f: error while loading shared libraries: libstd-b420905fdf897c89.so: cannot open shared object file: No such file or directory
```

The odd part is that particular project that fails has 0 tests, so I wonder if that is playing a part in it.

> musl

Tried that as well but didn't seem to work :( Though, do I need to specify `--target` when I install cargo-tarpaulin?

**Reply by xd009642** [reply]
Yeah the missing libstd and only libstdc++ will be the issue. And when installing with musl I think you do need to specify the musl target-triple for your system

**Reply by juliusl** [reply]
So after a lot of trial and error it appears that something rustup does fixes everything. In the same CI pipeline if I install rust w/ rustup, everything just works. It seems like LD_LIBRARY_PATH gets configured properly and that seems to be making the difference.

**Reply by xd009642** [reply]
Ah that's good, you should be able to figure out the changes rustup does and then gradually remove it if needed as well :pray:
